packages/contracts-bedrock/src/libraries/SuperRoot.sol

```diff
@@ -0,0 +1,165 @@
+                mstore(0x00, 0x10345c7c)
+                revert(0x1c, 0x04)
+            }
+
+            // Store the data offsets on the stack. Arrays of dynamically sized objects in solidity
+            // are comprised of pointers, so we must allocate two blocks of memory - one for the array
+            // of pointers, and one for the data itself.
+            let ptrArrayMemoryOffset := add(outputRoots, 0x20)
+            let outputRootsMemoryOffset := add(ptrArrayMemoryOffset, shr(0x01, remaining))
+
+            // Iteratively decode the output roots, checking that they are sorted in ascending order.
+            let lastChainId := 0x00
+            for { let i := 0x00 } lt(i, remaining) { i := add(i, 0x40) } {
+                let chainIdLocalOffset := i
+                let outputRootLocalOffset := add(i, 0x20)
+
+                // Extract the chain ID and output root commitment from the calldata.
+                let chainId := calldataload(add(outputRootsCalldataOffset, chainIdLocalOffset))
+                let outputRoot := calldataload(add(outputRootsCalldataOffset, outputRootLocalOffset))
+
+                // Check that the chain IDs are sorted in ascending order.
+                if lt(chainId, lastChainId) {
+                    // Signature of `OutputRootsNotSorted()` error
+                    mstore(0x00, 0x0277c843)
+                    revert(0x1c, 0x04)
+                }
+                lastChainId := chainId
+
+                // Store the tuple data.
+                let dataPtr := add(outputRootsMemoryOffset, i)
+                mstore(dataPtr, chainId)
+                mstore(add(dataPtr, 0x20), outputRoot)
+
+                // Store the pointer to the tuple data in the array of pointers.
+                mstore(add(ptrArrayMemoryOffset, shr(0x01, i)), dataPtr)
+            }
+
+            // Store the length of the output root tuple array in memory.
+            mstore(outputRoots, shr(0x06, remaining))
+
+            // Compute the new free memory pointer, adding the length of the output roots array
+            // (32 + num tuples * 32) and the length of the tuple data (num tuples * 64).
+            let endPtr := add(outputRoots, add(add(0x20, shr(0x01, remaining)), remaining))
+            let fmp := and(not(0x1F), add(endPtr, 0x1F))
+
+            // Finally, update the free memory pointer to the nearest aligned word to account for
+            // our allocations.
+            mstore(0x40, fmp)
+        }
+
+        superRoot_ = SuperRoot({ timestamp: rootTimestamp, outputRoots: outputRoots });
+    }
+
+    /// @notice Finds the output root with the given chain ID in the SuperRoot.
+    /// @param _superRoot The SuperRoot to search.
+    /// @param _chainId The chain ID to search for.
+    /// @return outputRoot_ The output root committing to the given chain ID at the timestamp
+    ///                     of the super root.
+    function findOutputRoot(
```

Comment:
We can remove the need to perform this search if we have the client pass more information to `proveWithdrawalTransaction`. They already have the super root locally, so they should be able to just pass the index within the array, and we can then just perform a quick sanity check.